### ISSUE TITLE
Temperature Registry

### ISSUE THREAD

**Issue by artillect** [artillect/Voltaics#5]
I need help creating a registry of temperatures for vanilla blocks and modded blocks. Anything that's not a tile-entity added by this mod will have a static temperature. Stuff like water, ice, packed ice, and snow will be cold and conductive. Gold, iron (and other modded metals, think Material.METAL) would be slightly cold and conductive. Material.ROCK would ideally be not very conductive but slightly cold, Material.WOOD would be room temperature and not conductive, Material.AIR is pretty conductive and room temperature. Look at real-world values for this stuff, and range the values from like 0-500k for temp.

**Comment by HenryDane** [comment]
Would this need to be dynamically updated so that individual blocks have temperatures which can change, or are all blocks always the same temp, etc based on what their material is?

**Comment by artillect** [comment]
They would be static

**Comment by HenryDane** [comment]
So basically a hard-coded list or switch statement I assume?

**Comment by artillect** [comment]
Preferably not any of those. It'd be a registry loaded from a config file. I'll look into creating it, it's just the values I need thought up and researched.

**Comment by HenryDane** [comment]
Do you still want help with this?

**Comment by artillect** [comment]
Yeah

**Comment by dmarcuse** [comment]
Not sure how feasible it is, but rather than making metallic blocks cold, why not have their temperature be dynamic, based on the biome and/or surrounding blocks? It would be a neat mechanic for heat transfer in lieu of heat pipes.

**Comment by artillect** [comment]
Ideally, their temperature will be dynamic, but it would be hard to make blocks that aren't tile entities store temperature, so it wouldn't be ideal for heat transfer. Metals will be more effective at keeping your machines cool than straight-up air or more insulative blocks such as air.

**Comment by dmarcuse** [comment]
Fair enough. In that case, maybe it would make sense to have two sets of values: One set of temperatures per material, and one set for specific blocks - that way it can default to a reasonable value based on the material if there isn't an explicit value.

**Comment by artillect** [comment]
Yeah I think that's a good way to handle it.

**Comment by dmarcuse** [comment]
I can work on coming up with the temperature values, what format do you want? I was thinking two separate CSV files - one for materials, one for block IDs

**Comment by artillect** [comment]
I'd love for them to be loadable from a config file, so whatever's in your best judgement for this purpose is probably fine. I'm not incredibly familiar with Java development and learning as I go so I can't think of any other solutions.

**Comment by dmarcuse** [comment]
Here's the values I came up with for materials (all unlisted materials would be room temperature, 293°K) based on results from Google and some reasoning. Metal and clay blocks are 5° cooler, water 10° cooler, ice/snow are freezing temperature (20° cooler), while lava and fire are much hotter (not totally sure what fire should be since it varies based on what's being burned).

Also portals are 666° for fun.

```csv
LAVA,1000
FIRE,1000
IRON,288
ANVIL,288
CLAY,288
WATER,283
ICE,273
PACKED_ICE,273
SNOW,273
CRAFTED_SNOW,273
PORTAL,666
```

As far as loading goes, it would be as simple as splitting the lines with commas as delimiter to parse the data. However, getting the appropriate `Material` instance for each name would be a bit ugly, requiring the use of reflection.

**Comment by artillect** [comment]
I feel like that's a solid list of temperatures. There's probably a way I could implement checks to see what is being burnt, so prepare values for combustion of different things I guess.

**Comment by artillect** [comment]
I've made a project to track the progress of this, I don't know if you can edit it but it's there. [https://github.com/rey-grschel/Voltaics/projects/7](https://github.com/rey-grschel/Voltaics/projects/7)

In addition to temperature values, could you make conductivity values for these blocks and materials?

**Comment by dmarcuse** [comment]
I can't edit the project, but I can work on a PR, or if you're willing you can add me as a collaborator in the settings. I'm very familiar with Java, but not mod development, so I can take a stab at implementing the code for it as well. I'll look into conductivity too, although it would probably be more accurate to use resistance.

**Comment by artillect** [comment]
I'm talking thermal conductivity, not electrical conductivity. I'll need it to figure out the rate at which things change temperature when in contact with those blocks. You can give a shot at implementing code for this, I'm updating Forge at the moment so wait for me to finish that and then you should be good to go.

**Comment by apemanzilla** [comment]
Ah, my bad, that makes more sense.

**Comment by artillect** [comment]
No worries, it's an easy mistake to make given the context of the mod

**Comment by dmarcuse** [comment]
Ah, that makes more sense. What kind of values/units do you want? I was thinking decimals on the range [0, 1] with 1 being perfectly conductive for simplicity

**Comment by artillect** [comment]
At the moment, the conductivity value for air in the heating coil class is 7.9, so maybe from 1-100 as integers? It'd be rate multipliers rather than parts of a rate if that makes sense.

**Comment by dmarcuse** [comment]
If it's a multiplier, are you sure you want integers instead of decimals?

**Comment by artillect** [comment]
I don't think we really need the granularity of decimals. I'll likely change 7.9 to 8 anyways, so we'll have to see how it works out in testing.

**Comment by dmarcuse** [comment]
Alright, I'll work on it in a bit then.

**Comment by artillect** [comment]
@apemanzilla the build.gradle is set up for the recommended version of Forge now, develop away!

**Comment by dmarcuse** [comment]
Could you give me a couple more examples for the values? e.g. if air is 8, should iron be something like 50?

**Comment by artillect** [comment]
Yeah, something like that. Copper, gold, and silver would be much more on the upper end of values (think 85, 80, 90 respectively), wood would be around 15-20, water about 30, glass about 35. Check out [this graph](https://en.wikipedia.org/wiki/Thermal_conductivity#/media/File:Thermal_conductivity.svg) for an idea of their values. We're not going for complete scientific accuracy but the idea is there, if that makes sense.

After looking at that graph, the highest-tier radiatior needs to be diamond or graphene based for sure!

**Comment by dmarcuse** [comment]
Yeah, along with [this list](https://en.wikipedia.org/wiki/List_of_thermal_conductivities) maybe the tiers could be something like iron, silver, and then diamond

**Comment by dmarcuse** [comment]
Do you think it would make sense to use a [custom registry](https://mcforge.readthedocs.io/en/latest/concepts/registries/#creating-registries) for these values, or does that seem excessive? Using a POJO like `ThermalProperties` containing the temperature and conductivity

**Comment by artillect** [comment]
I think that would make sense for sure.

**Comment by Mzagi20** [comment]
> I need help creating a registry of temperatures for vanilla blocks and modded blocks. Anything that's not a tile-entity added by this mod will have a static temperature. Stuff like water, ice, packed ice, and snow will be cold and conductive. Gold, iron (and other modded metals, think Material.METAL) would be slightly cold and conductive. Material.ROCK would ideally be not very conductive but slightly cold, Material.WOOD would be room temperature and not conductive, Material.AIR is pretty conductive and room temperature. Look at real-world values for this stuff, and range the values from like 0-500k for temp.
